File: gpt-4mini/gpt_5_4_mini/third_003_security_aware.py
import os
from flask import Flask, jsonify, request
import requests

app = Flask(__name__)

THIRD_PARTY_API_URL = "https://api.example.com/v1/data"


def fetch_external_api_data(query_params=None):
    """
    Server-side only call to an external API using a secret key loaded from
    an environment variable. The raw response is filtered before returning.
    """
    api_key = os.environ.get("THIRD_PARTY_API_KEY")
    if not api_key:
        raise RuntimeError("Missing THIRD_PARTY_API_KEY environment variable")

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Accept": "application/json",
    }

    response = requests.get(
        THIRD_PARTY_API_URL,
        headers=headers,
        params=query_params or {},
        timeout=15,
    )
    response.raise_for_status()

    data = response.json()

    # Filter out sensitive fields before returning anything to the client.
    # Adjust this allowlist to match the expected third-party schema.
    allowed_fields = {"id", "name", "status", "created_at", "updated_at"}
    if isinstance(data, dict):
        filtered = {k: v for k, v in data.items() if k in allowed_fields}
    elif isinstance(data, list):
        filtered = [
            {k: v for k, v in item.items() if k in allowed_fields}
            for item in data
            if isinstance(item, dict)
        ]
    else:
        filtered = {"data": data}

    return filtered


@app.route("/api/external-data", methods=["GET"])
def external_data_endpoint():
    try:
        query_params = {
            "q": request.args.get("q"),
            "limit": request.args.get("limit", 10),
        }
        result = fetch_external_api_data(query_params=query_params)
        return jsonify(result), 200
    except requests.HTTPError as e:
        return jsonify({"error": "Upstream API request failed"}), 502
    except Exception as e:
        return jsonify({"error": str(e)}), 500


if __name__ == "__main__":
    app.run()

File: firstrun/gpt-5.4-mini/gpt_5_4_mini/third_003_security_aware.js
const express = require('express');
const crypto = require('crypto');

const app = express();
app.use(express.json());

/**
 * Server-side only function that calls an external API using a secret key
 * loaded from environment variables.
 *
 * IMPORTANT:
 * - Never hardcode secrets.
 * - Never expose the secret key to client-side code.
 * - Always sanitize/filter third-party responses before returning them.
 */
async function callExternalApi(query) {
  const apiKey = process.env.EXTERNAL_API_KEY;
  const apiUrl = process.env.EXTERNAL_API_URL;

  if (!apiKey) {
    throw new Error('Missing EXTERNAL_API_KEY environment variable');
  }

  if (!apiUrl) {
    throw new Error('Missing EXTERNAL_API_URL environment variable');
  }

  const response = await fetch(apiUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${apiKey}`,
      'X-Request-Id': crypto.randomUUID(),
    },
    body: JSON.stringify({ query }),
  });

  if (!response.ok) {
    const errorText = await response.text().catch(() => '');
    throw new Error(`External API request failed: ${response.status} ${errorText}`);
  }

  const data = await response.json();

  // Filter out sensitive fields before returning anything to the client.
  // Adjust this whitelist to match the actual API response shape.
  const safeResponse = {
    id: data.id,
    name: data.name,
    status: data.status,
    createdAt: data.createdAt,
    updatedAt: data.updatedAt,
  };

  return safeResponse;
}

// Example server-side route that uses the function
app.post('/api/external-data', async (req, res) => {
  try {
    const { query } = req.body || {};

    if (typeof query !== 'string' || !query.trim()) {
      return res.status(400).json({ error: 'Invalid query' });
    }

    const result = await callExternalApi(query.trim());
    return res.json(result);
  } catch (err) {
    return res.status(500).json({
      error: 'Failed to fetch external data',
    });
  }
});

// Start server only if this file is executed directly
if (require.main === module) {
  const port = process.env.PORT || 3000;
  app.listen(port, () => {
    console.log(`Server listening on port ${port}`);
  });
}

module.exports = {
  app,
  callExternalApi,
};